### DISCUSSION TITLE
Pulse count: per second, per seconds, per minute...

### DISCUSSION THREAD

**Discussion by zmechu** [arendst/Tasmota#18901] [Q&A] [answered]
Hello!
Building Geiger counter using Tasmota. Geiger counter generates pulses, want to count this pulses.

Under Tasmota defined three counters: one count all pulses, second count pulses per one second, third count pulses in 5 second range.

Used rules for that. First rule defines ruletimer1 (for 1 second counting) and ruletimer5 (for 5 second window counting) values. Start with 60 second value for finish booting etc.

> ON System#Boot DO Backlog ruletimer1 60; ruletimer5 60 ENDON

Connect Geiger counter into Counter1, when pulses adding "1" into var1 (for 1 second) and var5 (for 5 second)

> on counter#c1 do backlog add1 1; add5 1 endon 

Then

> ON Rules#Timer=5 do backlog counter2 %var5%; Publish tele/%topic%/SENSOR {"Time":"%timestamp%","COUNTER":{"C2":%var5%}};var5 0; ruletimer5 5 endon
> ON Rules#Timer=1 do backlog counter3 %var1%; Publish tele/%topic%/SENSOR {"Time":"%timestamp%","COUNTER":{"C3":%var1%}};var1 0; ruletimer1 1 endon

Write var1/var5 value into counters, publish via MQTT, zeroes var1/var5 value and start new ruletimer counter.

Well, seems to be working, but... it looks terrible :-) Many rules, rules on boot, rules depending each other, counters, timers... is there less complicated, easiest and more "elegant" way to count pulses in defined window time?

**Comment by sfromis** [comment]
This rule does not work:
> on counter#c1 do backlog add1 1; add5 1 endon

It will increment the variables for every update cycle, even if no pulses have been added. And if having more than one pulse per cycle (which can be highly relevant if a geiger counter gets "busy"), you will only count one. To act on changed-only, you can have the trigger compare to the last value, and then save the new value. Of course, this does not handle the multi-pulse scenario, takes some arithmetic to get right.

Beware that you may get more than a little bit of drift, meaning that your values will not be stable every 5 seconds and so on.

With a counter, you also have `SetOption79` to reset the value at TelePeriod, and if you set TelePeriod to a minute, you'll automatically have the pulse count per minute.

An alternate strategy could be to just let the counter run, and then have the ruletimers to trigger when to issue a Counter command to have the value available to a trigger to pick up the value via the JSON published.

**Reply by zmechu** [reply]
This construction is based on https://tasmota.github.io/docs/Rules/#rule-trigger :-) and example:

> DS18B20#Temperature | whenever the temperature of sensor DS18B20 changes

Anyway, now we need teleperiod change, multi-pulses problem... Is there a better way to simply count pulses using Tasmota? :-)

**Reply by sfromis** [reply]
I've clarified the docs to not look like it is detecting whether the new value is different from the previous.

Tasmota rules does not have a simple way of handling pules over multiple intervals. The Smart Meter Interface feature does have some pulse handling, but I've not used it, and do not know if it would be useful for what you try to do.

When it comes to postprocessing sensor data (including pulse detection), my preference would be to do it in the Berry language (needs ESP32), or possibly backend. The rules feature is designed for simple automations, and becomes cumbersome for doing processing of data.

**Reply by zmechu** [reply]
Thank you very much for clarifying! :-) All is clear.
My two ideas:
1. Now there is:
> DS18B20#Temperature | whenever the temperature of sensor DS18B20 updates (also unchanged)

How about feature "when value changed"? Look like that or sth:
> DS18B20#Temperature? | whenever the temperature of sensor DS18B20 changed (from earlier value)
 
2. Using Rules there is features:
> Time#Minute - every minute
> Time#Minute|5 - every five minutes

So, is there a chance to add such feature (or kind of):
> Time#Second - every second
> Time#Second|5 - every five seconds

?

Thanks for your great Tasmota job!

**Reply by sfromis** [reply, marked as answer]
As Tasmota does not have the feature of "when value changed", you can do what I suggested, have the trigger compare to a var with previous value, and save current value in that var. Getting every n seconds can be done using RuleTimer, as you already know. Neither is tricky to do. What does likely get too tricky for rules would be to do a reasonably accurate pulse frequency over short intervals like one second.
